python/tank/templatekey.py

```diff
@@ -256,6 +273,145 @@ def _as_string(self, value):
+        Constructor
+        :param name: Name by which the key will be refered.
+        :param utc_default: If True, the default value will be an UTC time. Defaults to False.
+        :param format_spec: Specification for formating when casting to/from a string.
+                            The format follows the convention of strftime and strptime. The
+                            default value is "%Y-%m-%d-%H-%M-%S". Given June 24th, 2015 at
+                            9:20:30 PM, this will yield 2015-06-24-21-20-30
+        """
+        if isinstance(format_spec, basestring) is False:
+            raise TankError("format_spec for TimestampKey %s is not of type string: %s" %
+                            (name, format_spec.__class__.__name__))
+        self.format_spec = format_spec
+
+        if isinstance(utc_default, bool) is False:
+            raise TankError("utc_default for TimestampKey %s is not of type boolean: %s" %
+                            (name, utc_default.__class__.__name__))
+        self.utc_default = utc_default
+
+        super(TimestampKey, self).__init__(
+            name,
+            default=self.__get_current_time
+        )
+
+    def __get_current_time(self):
+        """
+        Returns the current time as a datetime.datetime instance.
+
+        Do not streamline the code so the __init__ method simply passesd the dt.datetime.now method,
+        we can't mock datetime.now since it's builtin and will make unit tests more complicated to
+        write.
+
+        :returns: If utc_default, a datetime object representing time in the UTC timezone. Otherwise,
+                  a datetime object representing time in the local timezone. In any case, the tzinfo
+                  member is None.
+        """
+        if self.utc_default:
+            return dt.datetime.utcnow()
+        else:
+            return dt.datetime.now()
+
+    def validate(self, value):
+        """
+        Test if a value is valid for this key.
+
+        :param value: Value to test.
+
+        :returns: Bool
+        """
+        if isinstance(value, basestring) or isinstance(value, unicode):
+            return self._validate_str(value)
+        else:
+            return self._validate_value(value)
+
+    def _validate_str(self, str_value):
+        """
+        Test if a string is valid for this key.
+
+        :param str_value: String to test.
+
+        :returns: Bool
+        """
+        try:
+            self._as_value(str_value)
+            return True
+        except ValueError, e:
+            self._last_error = "Invalid string: %s" % e.message
+            return False
+        except TypeError, e:
+            self._last_error = "Invalid type: %s" % e.message
+            return False
+
+    def _validate_value(self, value):
+        """
+        Test if a string is valid for this key.
+
+        :param str_value: String to test.
+
+        :returns: Bool
+        """
+        try:
+            self._as_string(value)
```

Comment:
_as_string is a method from the base class that needs to be overriden, I can't just inline it. ;)

The reason why I as calling _as_string and _as_value in the _validate_* methods is that the actual validation of the input value can only be done by trying to convert the string to a date and vice versa, since the strptime and strftime methods are the ones who know what are the error conditions.